Added Contributors Guide (#165)
ansible-collections/dellemc.enterprise_sonic: docs/CONTRIBUTING.md

<!--
Copyright (c) 2022 Dell Inc., or its subsidiaries. All Rights Reserved.

Licensed under the GPL, Version 3.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://www.gnu.org/licenses/gpl-3.0.txt
-->

# How to Contribute

> Become one of the contributors to this project! We thrive to build a welcoming and open community for anyone who wants to use the project or contribute to it. There are just a few small guidelines you need to follow. To help us create a safe and positive community experience for all, we require all participants to adhere to the [Code of Conduct](CODE_OF_CONDUCT.md).

Author: addressed as suggested.

The owning team _squashes_ all commits into one when we accept a pull request. The title of the pull request becomes the subject line of the squashed commit message. We still encourage contributors to write informative commit messages, as they become a part of the Git commit body.

We use the pull request title when we generate change logs for releases. As such, we strive to make the title as informative as possible.

Make sure that the title for your pull request uses the same format as the subject line in the commit message.

### Quality Gates for pull requests

GitHub Actions are used to enforce quality gates when a pull request is created or when any commit is made to the pull request. These GitHub Actions enforce our minimum code quality requirement for any code that gets checked into the code repository. If any of the quality gates fail, it is expected that the contributor will look into the check log, understand the problem and resolve the issue. If help is needed, please feel free to reach out to the maintainers of the project for [support](SUPPORT.md).

#### Code build/test/coverage

> [GitHub action](https://github.com/dell/<<Action link>>) that runs unit tests and checks that the code coverage of each package meets a configured threshold (currently 90%). An error is flagged if a given pull request does not meet the test coverage threshold and blocks the pull request from being merged.

Author: We have updated the section as "[GitHub action](https://github.com/ansible-collections/dellemc.enterprise_sonic/actions) that runs unit tests automatically and checks that the code coverage of each package meets a configured threshold (currently 90%). An error is flagged if a given pull request does not meet the test coverage threshold and blocks the pull request from being merged. When it fails, it is expected that the contributor will look into the log, understand the problem and resolve the issue."

> [GitHub action](https://github.com/dell/<<Action link>>) that runs unit tests and checks that the code coverage of each package meets a configured threshold (currently 90%). An error is flagged if a given pull request does not meet the test coverage threshold and blocks the pull request from being merged.

Author: Updated as suggested.

> [GitHub action](https://github.com/dell/<<Action link>>) that runs unit tests and checks that the code coverage of each package meets a configured threshold (currently 90%). An error is flagged if a given pull request does not meet the test coverage threshold and blocks the pull request from being merged.

Reviewer: This is a better description, but I still have a couple of minor comments. Since I have posted them as a comment in the file (with a specific request for changed content), I will mark this issue "Resolved" to simplify tracking of the remaining issues.

## Code Reviews

All submissions, including submissions by project members, require review. We use GitHub pull requests for this purpose. Consult [GitHub Help](https://help.github.com/articles/about-pull-requests/) for more information on using pull requests.

A pull request must satisfy the following for it to be merged:

* A pull request will require at least 2 maintainer approvals.
* Maintainers must perform a review to ensure the changes adhere to guidelines laid out in this document.
* If any commits are made after the PR has been approved, the PR approval will automatically be removed and the above process must happen again.

## Code Style

> For the Python code in the repository, we expect the code styling outlined in [Ansible python guide]( https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html). In addition to this, we have the following supplements:

Author: Updated the section.

> ### Handle Errors

Author: removed as suggested.

> ### Handle Errors

Reviewer: Verified.

An Issue __must__ be created before submitting any pull request. Any pull request that is created should be linked to an Issue.

### Feature request

If you have an idea of how to improve this project, submit a [feature request](https://github.com/dell/<<repo>>/issues/new?labels=type%2Ffeature-request%2C+needs-triage&template=feature_request.md&title=%5BFEATURE%5D%3A).

### Answering questions

If you have a question and you can't find the answer in the documentation or issues, the next step is to submit a [question.](https://github.com/dell/<<repo>>/issues/new?labels=type%2Fquestion&template=ask-a-question.md&title=%5BQUESTION%5D%3A)

We'd love your help in answering pending questions asked by other CSM users.

## Triage issues

Triage helps ensure that issues resolve quickly by:

* Ensuring the issue's intent and purpose is conveyed precisely. This is necessary because it can be difficult for an issue to explain how an end user experiences a problem and what actions they took.
* Giving a contributor the information they need before they commit to resolving an issue.

> * Lowering the issue count by preventing duplicate issues.

Reviewer 2: ```suggestion

```

Author: removed as suggested

Reviewer: This is another change that apparently was made and was subsequently reversed. I am unresolving this issue.

Please figure out how this happened to prevent future similar recurrences.

To resolve this issue, either remove this line or modify it as suggested in this issue.

Author: addressed as suggested

> * Streamlining the development process by preventing duplicate discussions.

Reviewer 2: ```suggestion
* Streamlining the development process by identifying and closing duplicate issues.
```

Author: Addressed as suggested

Author: Removed as suggested.

Reviewer: Verified. 

Note: There is still an issue associated with this line, but I'm closing this one to simplify tracking of the remaining issues.

* [Help wanted](https://github.com/dell/<<repo>>/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22) issues are problems we would like the community to help us with regardless of complexity.

When you're ready to contribute, it's time to create a pull request.

## Branching

* [Branching Strategy](BRANCHING.md) 

## Signing your commits

We require that developers sign off their commits to certify that they have permission to contribute the code in a pull request. This way of certifying is commonly known as the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). We encourage all contributors to read the DCO text before signing a commit and making contributions.

GitHub will prevent a pull request from being merged if there are any unsigned commits.

### Signing a commit

GPG (GNU Privacy Guard) will be used to sign commits.  Follow the instructions [here](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/signing-commits) to create a GPG key and configure your GitHub account to use that key.

> Make sure you have your user name and e-mail set.  This will be required for your signed commit to be properly verified.  Check the following references:

Reviewer 2: ```suggestion
Make sure you have your username and e-mail set.  This will be required for your signed commit to be properly verified.  Check the following references:
```

Author: Addressed as suggested

Author: The change has been incorporated.

Reviewer: Verified.

> * Setting up your github user name [reference](https://help.github.com/articles/setting-your-username-in-git/)

Reviewer 2: ```suggestion
* Setting up your github username [reference](https://help.github.com/articles/setting-your-username-in-git/)
```

Author: Addressed as suggested

Author: The change has been incorporated.

Reviewer: Verified.

<!--
Copyright (c) 2022 Dell Inc., or its subsidiaries. All Rights Reserved.

Licensed under the GPL, Version 3.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://www.gnu.org/licenses/gpl-3.0.txt
-->

# How to Contribute

Become one of the contributors to this project! We strive to build a welcoming and open community for anyone who wants to use the project or contribute to it. There are just a few small guidelines you need to follow. To help us create a safe and positive community experience for all, we require all participants to adhere to the [Code of Conduct](CODE_OF_CONDUCT.md).

## Table of Contents

* [Become a contributor](#Become-a-contributor)
* [Contributor agreement](#Contributor-Agreement)
* [Submitting issues](#Submitting-issues)
* [Triage issues](#Triage-issues)
* [Your first contribution](#Your-first-contribution)
* [Branching](#Branching)
* [Signing your commits](#Signing-your-commits)
* [Pull requests](#Pull-requests)
* [Code reviews](#Code-reviews)

> * [TODOs in the code](#TODOs-in-the-code)

Author: The links are directed within the document.

> * [TODOs in the code](#TODOs-in-the-code)

Reviewer: This section was replaced or renamed by "Code Style", but the link from the table of contents doesn't work. I have posted an inline comment for the new line.

> * [TODOs in the code](#TODOs-in-the-code)

Author: addressed as suggested

> * [TODOs in the code](#TODOs-in-the-code)

Reviewer: Verified.

We'd love your help in answering pending questions asked by other CSM users.

## Triage issues

Triage helps ensure that issues resolve quickly by:

* Ensuring the issue's intent and purpose is conveyed precisely. This is necessary because it can be difficult for an issue to explain how an end user experiences a problem and what actions they took.
* Giving a contributor the information they need before they commit to resolving an issue.
* Lowering the issue count by preventing duplicate issues.
* Streamlining the development process by preventing duplicate discussions.

If you don't have the knowledge or time to code, consider helping with _issue triage_. The Dell dellemc.enterprise_sonic community will thank you for saving them time by spending some of yours.

Read more about the ways you can [Triage issues](ISSUE_TRIAGE.md).

## Testing

See [here](https://github.com/ansible-collections/dellemc.enterprise_sonic/tree/collections/README.md) for further information on testing.

> ## Debugging

Author: Addressed as suggested.

> ## Debugging

Reviewer: This has not been addressed. (Or, if it was, the change got accidentally backed out.) Please make (or re-make) the requested change to resolve this issue.

> ## Debugging

Author: addressed as suggested

To debug <product> using IDE, see [here](https://github.com/ansible-collections/dellemc.enterprise_sonic/tree/collections/docs/debug.md)

## Your first contribution

Unsure where to begin contributing? Start by browsing issues labeled `beginner friendly` or `help wanted`.

* [Beginner-friendly](https://github.com/ansible-collections/dellemc.enterprise_sonic/issues?q=is%3Aopen+is%3Aissue+label%3A%22beginner+friendly%22) issues are generally straightforward to complete.
* [Help wanted](https://github.com/ansible-collections/dellemc.enterprise_sonic/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22) issues are problems we would like the community to help us with regardless of complexity.

When you're ready to contribute, it's time to create a pull request.

## Branching

> * [Branching Strategy](BRANCHING.md) 

Author: Have created a new file for branching file as like open manage. We will update the file once the server storage team updates the file in open manage.

> * [Branching Strategy](BRANCHING.md) 

Reviewer: Okay.

* Title and description matches the implementation.
* Commits within the pull request follow the formatting guidelines.
* The pull request closes one related issue.
* The pull request contains necessary tests that verify the intended behavior.
* If your pull request has conflicts, rebase your branch onto the main branch.

If the pull request fixes a bug:

* The pull request description must include `Fixes #<issue number>`.
* To avoid regressions, the pull request should include tests that replicate the fixed bug.

The owning team _squashes_ all commits into one when we accept a pull request. The title of the pull request becomes the subject line of the squashed commit message. We still encourage contributors to write informative commit messages, as they become a part of the Git commit body.

We use the pull request title when we generate change logs for releases. As such, we strive to make the title as informative as possible.

Make sure that the title for your pull request uses the same format as the subject line in the commit message.

### Quality Gates for pull requests

> GitHub Actions are used to enforce quality gates when a pull request is created or when any commit is made to the pull request. These GitHub Actions enforce our minimum code quality requirement for any code that gets checked into the code repository. If any of the quality gates fail, it is expected that the contributor will look into the check log, understand the problem and resolve the issue. If help is needed, please feel free to reach out to the maintainers of the project for [support](SUPPORT.md).

Reviewer: This link doesn't exist. Please either add a little content inline here instead of an empty link.

(Or just omit the link and end the sentence with the word "support" (not as a link).

Author: Have updated the merged link for support

Reviewer: Verified.

#### Code build/test/coverage

> [GitHub action](https://github.com/ansible-collections/dellemc.enterprise_sonic/actions) that runs unit tests automatically and checks that the code coverage of each package meets a configured threshold (currently 90%). An error is flagged if a given pull request does not meet the test coverage threshold and blocks the pull request from being merged. When it fails, it is expected that the contributor will look into the log, understand the problem and resolve the issue.  

Reviewer: We have a real workflow available to reference for this now. Please use that instead of the link provided here.

Also, please update this sentence to make it grammatical. (An example is shown below.)

*********************
e.g.

The code coverage tool, [code-coverage.yml](https://github.com/ansible-collections/dellemc.enterprise_sonic/blob/main/.github/workflows/code-coverage.yml), runs unit tests automatically and checks that the code coverage of each package meets a configured threshold (currently 90%)
*********************

 Please also add a sentence at the end of or after this paragraph providing instructions for running the code coverage test. Arun can provide you with the syntax for this.

Author: The change has been incorporated as suggested. 
__________________________________________________________________

You had asked us to provide the instructions to run the code coverage test.Do you mean to say that we shall add the following lines ? We have updated the below lines in the file as suggested. 

Alternatively, users can manually run the unit test and check the coverage using 'ansible-test' command as given in the following example. 

```
ansible-test  units --color --python 3.10 --coverage -vvvv tests/unit/modules/network/sonic/test_sonic_bgp_neighbors.py
ansible-test coverage report

Reviewer: Yes. Thank you. I have added an inline comment requesting a couple of minor changes to the added lines, so I will mark this issue "Resolved" to facilitate tracking.

## Code Reviews

All submissions, including submissions by project members, require review. We use GitHub pull requests for this purpose. Consult [GitHub Help](https://help.github.com/articles/about-pull-requests/) for more information on using pull requests.

A pull request must satisfy the following for it to be merged:

* A pull request will require at least 2 maintainer approvals.
* Maintainers must perform a review to ensure the changes adhere to guidelines laid out in this document.
* If any commits are made after the PR has been approved, the PR approval will automatically be removed and the above process must happen again.

## Code Style

For the Python code in the repository, we expect the code styling outlined in [Ansible python guide]( https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html). In addition to this, we have the following supplements:

* Contributions should adhere to ansible Coding standard guidelines as we follow these standards.

> * Should include [test](https://github.com/ansible-collections/dellemc.enterprise_sonic/tree/collections/tests) scripts for your changes.

Reviewer: This link is incorrect. Instead, please use the following link if it points to the intended location:

https://github.com/ansible-collections/dellemc.enterprise_sonic/tree/main/tests

<!--
Copyright (c) 2022 Dell Inc., or its subsidiaries. All Rights Reserved.

Licensed under the GPL, Version 3.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://www.gnu.org/licenses/gpl-3.0.txt
-->

# How to Contribute

Become one of the contributors to this project! We strive to build a welcoming and open community for anyone who wants to use the project or contribute to it. There are just a few small guidelines you need to follow. To help us create a safe and positive community experience for all, we require all participants to adhere to the [Code of Conduct](CODE_OF_CONDUCT.md).

## Table of Contents

* [Become a contributor](#Become-a-contributor)
* [Contributor agreement](#Contributor-Agreement)
* [Submitting issues](#Submitting-issues)
* [Triage issues](#Triage-issues)
* [Your first contribution](#Your-first-contribution)
* [Branching](#Branching)
* [Signing your commits](#Signing-your-commits)
* [Pull requests](#Pull-requests)
* [Code reviews](#Code-reviews)

> * [Code Style](#Code Style)

Reviewer: This link doesn't work. Please add a hypen between "Code" and "Style" to fix it.

Author: addressed as suggested

> * [Code Style](#Code Style)

Reviewer: ```suggestion
* [Code Style](#Code-Style)
```

Author: addressed as suggested

Reviewer: Verified.

#### Code build/test/coverage

[GitHub action](https://github.com/ansible-collections/dellemc.enterprise_sonic/actions) that runs unit tests automatically and checks the code coverage tool, [code-coverage.yml](https://github.com/ansible-collections/dellemc.enterprise_sonic/blob/main/.github/workflows/code-coverage.yml), runs unit tests automatically and checks that the code coverage of each package meets a configured threshold (currently 90%). An error is flagged if a given pull request does not meet the test coverage threshold and blocks the pull request from being merged. When it fails, it is expected that the contributor will look into the log, understand the problem and resolve the issue.  

> Alternatively, users can manually run the unit test and check the coverage using 'ansible-test' command as given in the following example. 

Reviewer: ```suggestion
Alternatively, users can manually run the unit test and check the coverage using the 'ansible-test' command as given in the following example. These commands must be run from the Ansible collection "install" directory for the enterprise sonic collection (e.g. "cd ~/.ansible/collections/ansible_collections/dellemc/enterprise_sonic").
```

Author: updated as suggested
